lib/SILOptimizer/FunctionSignatureTransforms/ExistentialSpecializer.cpp

```diff
@@ -64,12 +74,50 @@ class ExistentialSpecializer : public SILFunctionTransform {
     /// Get CallerAnalysis information handy.
     CA = PM->getAnalysis<CallerAnalysis>();
 
+    /// Get ProtocolConformanceAnalysis.
+    PCA = PM->getAnalysis<ProtocolConformanceAnalysis>();
+
+    /// Get ClassHierarchyAnalysis.
+    CHA = PM->getAnalysis<ClassHierarchyAnalysis>();
     /// Perform specialization.
     specializeExistentialArgsInAppliesWithinFunction(*F);
   }
 };
 } // namespace
 
+/// Find concrete type from Sole Conforming Type.
+bool ExistentialSpecializer::findConcreteTypeFromSoleConformingType(
+    SILFunctionArgument *Arg, CanType &ConcreteType) {
+  auto ArgType = Arg->getType();
+  auto SwiftArgType = ArgType.getASTType();
+
+  /// Do not handle composition types yet.
+  if (isa<ProtocolCompositionType>(SwiftArgType))
+    return false;
+  assert(ArgType.isExistentialType());
+  /// Find the protocol decl.
+  auto *PD = dyn_cast<ProtocolDecl>(SwiftArgType->getAnyNominal());
+  if (!PD)
+    return false;
+
+  // Determine the sole conforming type.
+  auto *NTD = PCA->findSoleConformingType(PD);
+  if (!NTD)
+    return false;
+
+  // Sole conforming class should not be open access or have any derived class.
+  ClassDecl *CD;
+  if ((CD = dyn_cast<ClassDecl>(NTD)) &&
+      (CD->getEffectiveAccess() == AccessLevel::Open ||
+       CHA->hasKnownDirectSubclasses(CD))) {
+    return false;
+  }
```

Comment:
This should be a standalone API because it has special logic to reason about possible conformances. That logic should only exist in one place in the compiler.

Reply:
Originally I had this in findSoleConformingType function of ProtocolConformanceAnalysis, however since this also required ClassHierarchyAnalysis, we had split. I can put this in Existential.cpp in Utils. This would be a wrapper for findSoleConformingType.

Reply:
I'm just trying to avoid duplicating logic that reasons about the type system in multiple places in the SILOptimizer. As an alternative, we could just have a `CHA->mayHaveSubclasses` query.